architecture/core.rst

```diff
@@ -174,98 +176,173 @@ with IRQ and FIQ blocked. SMCs are categorised in two flavors: **fast** and
-    :figclass: align-center
-    
-    IRQ received in secure world and forwarded to normal world
+.. uml::
+    :align: center
+    :caption: Foreign interrupt received in secure world and forwarded to
+              normal world
 
-**Deliver a non-secure interrupt to normal world when ``SCR_NS`` is set**
+    participant "Normal World" as nwd
+    participant "Secure Monitor" as smon
+    participant "OP-TEE OS entry" as entry
+    participant "OP-TEE OS" as optee
+    == IRQ and FIQ unmasked ==
+    optee -> optee : process
+    == IRQ and FIQ unmasked,\nForeign interrupt received ==
+    optee -> optee : suspend thread
+    optee -> entry : forward foreign interrupt
+    entry -> smon : smc: forward foreign interrupt
+    smon -> smon: Save secure context
+    smon -> smon: Restore non-secure context
+    == IRQ and FIQ unmasked ==
+    smon --> nwd : eret: IRQ forwarded
+    == FIQ unmasked, IRQ received ==
+    nwd -> nwd : process IRQ
+    == IRQ and FIQ unmasked ==
+    nwd -> smon : smc: return from IRQ
+    == IRQ and FIQ masked ==
+    smon -> smon : Save non-secure context
+    smon -> smon : Restore secure context
+    smon --> entry : eret: return from foreign interrupt
+    entry -> entry : find thread
+    entry --> optee : resume execution
+    == IRQ and FIQ unmasked ==
+    optee -> optee : process
 
-Since ``SCR_IRQ`` is cleared, an IRQ will be delivered using the state vector
-(``VBAR``) in the normal world. The IRQ is received as any other exception by
-normal world, the monitor and the Trusted OS are not involved at all.
+**Deliver a foreign interrupt to normal world when ``SCR_NS`` is set**
+
+Since ``SCR_IRQ`` is cleared, an IRQ will be delivered using the exception
+vector (``VBAR``) in the normal world. The IRQ is received as any other
+exception by normal world, the monitor and the OP-TEE OS are not involved
+at all.
 
 Deliver secure interrupts to Secure World
 =========================================
-This section uses the Arm GICv1/v2 conventions: FIQ signals secure interrupts
-while IRQ signals non-secure interrupts. On a GICv3 configuration, one should
-exchange IRQ and FIQ in this section. A FIQ can be received during two different
-states, either in normal world (``SCR_NS`` is set) or in secure world
-(``SCR_NS`` is cleared). When the secure monitor is active (Armv8-A EL3 or
-Armv7-A Monitor mode) FIQ is masked. FIQ reception in the two different states
-is described below.
+A secure (foreign) interrupt can be received during two different states,
+either in normal world (``SCR_NS`` is set) or in secure world (``SCR_NS``
+is cleared). When the secure monitor is active (Armv8-A EL3 or Armv7-A
+Monitor mode) FIQ and IRQ are masked. FIQ reception in the two different
+states is described below.
 
-**Deliver FIQ to secure world when SCR_NS is set**
+**Deliver secure interrupt to secure world when SCR_NS is set**
 
-When the monitor gets an FIQ exception it:
+When the monitor traps an secure interrupt exception it:
```

Comment:
I wonder if this sentence is wrong. I find it hard to read. It's the end that feels wrong `... interrupt exception it` `s/exception/handles/`?

Reply:
How about just:
> When the monitor traps a secure interrupt it:

Reply:
+1 Much better!